### PR TITLE
Rework v3 fuzzing

### PR DESCRIPTION
**Pull request by RumovZ**
Details in https://github.com/ankitects/anki/issues/1416#issuecomment-958208149.

How should I proceed with the Python tests? Do we stop testing v3 there and add Rust unit tests instead? I'll add some more comments in the code.

### CODE REVIEW

File: rslib/src/scheduler/states/mod.rs [hunk 1]
```diff
@@ -95,45 +95,62 @@ pub(crate) struct StateContext<'a> {
...
+            let (lower, upper) = constrained_fuzz_bounds(interval, minimum, maximum);
+            (lower as f32 + fuzz_factor * ((1 + upper - lower) as f32)).floor() as u32
         } else {
-            interval
+            (interval.round() as u32).max(minimum).min(maximum)
         }
-        .round() as u32
     }
 
     pub(crate) fn fuzzed_graduating_interval_good(&self) -> u32 {
```
Comment (RumovZ) [inline rslib/src/scheduler/states/mod.rs:118]: Probably doesn't matter in practice, but these respect the configured limits now.

File: rslib/src/scheduler/states/mod.rs [hunk 2]
```diff
@@ -95,45 +95,62 @@ pub(crate) struct StateContext<'a> {
...
+        fuzz_range(interval, 0.15, 2.0)
+    } else {
+        fuzz_range(interval, 0.05, 4.0)
     }
 }
 
-fn fuzz_range(interval: f32, factor: f32, minimum: f32) -> (f32, f32) {
+fn fuzz_range(interval: f32, factor: f32, minimum: f32) -> (u32, u32) {
     let delta = (interval * factor).max(minimum).max(1.0);
-    (interval - delta, interval + delta + 1.0)
```
Comment (RumovZ) [inline rslib/src/scheduler/states/mod.rs:136]: I couldn't figure out the purpose of the `+ 1.0`. Was it fine to drop it?

File: rslib/src/scheduler/answering/mod.rs [hunk 3]
```diff
@@ -59,7 +61,7 @@ impl CardStateUpdater {
     /// state handling code from the rest of the Anki codebase.
     pub(crate) fn state_context(&self) -> StateContext<'_> {
         StateContext {
-            fuzz_seed: self.fuzz_seed,
+            fuzz_factor: get_fuzz_factor(self.fuzz_seed),
```
Comment (RumovZ) [inline rslib/src/scheduler/answering/mod.rs:64]: I (ab?)used this to work around the different algorithm in `with_learning_fuzz()`.

File: rslib/src/scheduler/states/mod.rs (hunk 2 above)
Comment (dae) [reply rslib/src/scheduler/states/mod.rs:136]: I don't remember why I added that, and removing it from HEAD doesn't appear to break any of the tests, so it seems like it's safe to cut.

File: rslib/src/scheduler/states/mod.rs [hunk 4]
```diff
@@ -69,7 +68,8 @@ impl CardState {
 
 /// Info required during state transitions.
 pub(crate) struct StateContext<'a> {
-    pub fuzz_seed: Option<u64>,
+    /// In `[0, 1)`. Used to pick the final interval from the fuzz range.
```
Comment (dae) [inline rslib/src/scheduler/states/mod.rs:71]:
```suggestion
    /// Between 0.0-1.0. Used to pick the final interval from the fuzz range.
```

**Review by dae — COMMENTED** [review]
Looking good! I've pushed the beginnings of a unit test - if you change scheduled/elapesd days from 2 to 1, it will trigger a panic which needs looking into.

**Comment by dae** [comment]
The Python tests are run with fuzzing disabled, so theoretically this change shouldn't break the existing tests. Any ideas why reviewing_early_normal() is a day off now?

We can test the fuzzing code in Rust, and some tests there would be great.

**Comment by RumovZ** [comment]
> Any ideas why reviewing_early_normal() is a day off now?

I've removed flooring in 93708c6, because I thought my subsequent commits would break these tests anyway. Should I try to revert it?

File: rslib/src/scheduler/states/mod.rs (hunk 4 above)
Comment (RumovZ) [reply rslib/src/scheduler/states/mod.rs:71]: I wanted to emphasise that the upper bound is not included, but I can use more intelligible phrasing.

**Comment by RumovZ** [comment]
> I've pushed the beginnings of a unit test

Sorry, I couldn't find where.

**Comment by dae** [comment]
Sorry, I have a git commit hook that runs checks before pushing, and it failed because of the failing Python test. Should be available now: https://github.com/ankitects/anki/pull/1474/commits/4a745877846d29ec9e722b155d6d07424f67f17d

If it's only that one Python test we're breaking, then removing the flooring sounds good. It would be good if we could get some coverage of the same area either in Rust, or by cloning the Python function and making the minor changes required (that may be the more expedient option)

**Comment by RumovZ** [comment]
> if you change scheduled/elapesd days from 2 to 1, it will trigger a panic which needs looking into.

A fuzz factor of 1 is undefined behaviour as we sample with `gen_range(0.0..1.0)`. Not sure if we can do anything to prevent future mistakes apart from the doc string.

File: rslib/src/scheduler/states/review.rs [hunk 5]
```diff
@@ -253,4 +253,20 @@ mod test {
         assert!(leech_threshold_met(2, 1));
         assert!(leech_threshold_met(3, 1));
     }
+
+    #[test]
+    fn fuzz() {
```
Comment (RumovZ) [inline rslib/src/scheduler/states/review.rs:258]: I've added tests in `scheduler/states/mod.rs`. Do we still need tests here? I guess we could assert that intervals grow both for increasing fuzz factor and button numbers.

File: pylib/tests/test_schedv2.py [hunk 6]
```diff
@@ -715,6 +717,7 @@ def test_suspend():
 
 
 def test_filt_reviewing_early_normal():
+    to_int = round if is_2021() else int
```
Comment (dae) [inline pylib/tests/test_schedv2.py:720]: Nice simple solution :-)

File: rslib/src/scheduler/states/review.rs (hunk 5 above)
Comment (dae) [reply rslib/src/scheduler/states/review.rs:258]: Your tests look good, and this was just intended as a starting point; feel free to cut it.

File: rslib/src/scheduler/states/mod.rs (hunk 4 above)
Comment (dae) [reply rslib/src/scheduler/states/mod.rs]: "In range 0.0..1.0"? Or we could use the more explicit Swift syntax:  Between 0.0..<1.0

**Comment by dae** [comment]
> A fuzz factor of 1 is undefined behaviour as we sample with gen_range(0.0..1.0). Not sure if we can do anything to prevent future mistakes apart from the doc string. 

Ah, so it was user error on my part :-)

**Review by dae — COMMENTED** [review]
Looks good Rumo, apart from tweaking that comment and stripping the test in review.rs, I can't see anything that might need changing

**Comment by RumovZ** [comment]
Great, ready from my side as well. 🙂

**Comment by dae** [comment]
Thanks for this Rumo :-)

**Pull request merged**